storage/etcd/etcd.go

```diff
@@ -0,0 +1,472 @@
+		if now.After(authRequest.Expiry) {
+			if err := c.DeleteAuthRequest(authRequest.ID); err != nil {
+				c.logger.Errorf("failed to delete auth request: %v", err)
+				delErr = fmt.Errorf("failed to delete auth request: %v", err)
+			}
+			result.AuthRequests++
+		}
+	}
+	if delErr != nil {
+		return result, delErr
+	}
+
+	authCodes, err := c.listAuthCodes()
+	if err != nil {
+		return result, err
+	}
+
+	for _, authCode := range authCodes {
+		if now.After(authCode.Expiry) {
+			if err := c.DeleteAuthCode(authCode.ID); err != nil {
+				c.logger.Errorf("failed to delete auth code %v", err)
+				delErr = fmt.Errorf("failed to delete auth code: %v", err)
+			}
+			result.AuthCodes++
+		}
+	}
+	return result, delErr
+}
+
+func (c *conn) CreateAuthRequest(a storage.AuthRequest) error {
+	return c.txnCreate(context.TODO(), authRequestPrefix+a.ID, fromStorageAuthRequest(a))
+}
+
+func (c *conn) GetAuthRequest(id string) (a storage.AuthRequest, err error) {
+	var req AuthRequest
+	key := authRequestPrefix + id
+	if err = c.getOne(key, &req); err != nil {
+		return
+	}
+	return toStorageAuthRequest(req), nil
+}
+
+func (c *conn) UpdateAuthRequest(id string, updater func(a storage.AuthRequest) (storage.AuthRequest, error)) error {
+	key := authRequestPrefix + id
+	ctx := context.TODO()
+	return c.txnUpdate(ctx, key, func(currentValue []byte) ([]byte, error) {
+		var current AuthRequest
+		if len(currentValue) > 0 {
+			if err := json.Unmarshal(currentValue, &current); err != nil {
+				return nil, err
+			}
+		}
+		updated, err := updater(toStorageAuthRequest(current))
+		if err != nil {
+			return nil, err
+		}
+		return json.Marshal(fromStorageAuthRequest(updated))
+	})
+}
+
+func (c *conn) DeleteAuthRequest(id string) error { return c.deleteKey(authRequestPrefix, id) }
+
+func (c *conn) CreateAuthCode(a storage.AuthCode) error {
+	return c.txnCreate(context.TODO(), authCodePrefix+a.ID, fromStorageAuthCode(a))
+}
+
+func (c *conn) GetAuthCode(id string) (a storage.AuthCode, err error) {
+	key := authCodePrefix + id
+	err = c.getOne(key, &a)
+	return a, err
+}
+
+func (c *conn) DeleteAuthCode(id string) error { return c.deleteKey(authCodePrefix, id) }
+
+func (c *conn) CreateRefresh(r storage.RefreshToken) error {
+	return c.txnCreate(context.TODO(), refreshTokenPrefix+r.ID, fromStorageRefreshToken(r))
+}
+
+func (c *conn) GetRefresh(id string) (r storage.RefreshToken, err error) {
+	var token RefreshToken
+	key := refreshTokenPrefix + id
+	if err = c.getOne(key, &token); err != nil {
+		return
+	}
+	return toStorageRefreshToken(token), nil
+}
+
+func (c *conn) UpdateRefreshToken(id string, updater func(old storage.RefreshToken) (storage.RefreshToken, error)) error {
+	key := refreshTokenPrefix + id
+	ctx := context.TODO()
+	return c.txnUpdate(ctx, key, func(currentValue []byte) ([]byte, error) {
+		var current RefreshToken
+		if len(currentValue) > 0 {
+			if err := json.Unmarshal([]byte(currentValue), &current); err != nil {
```

Comment:
Any thoughts on using a faster serialization format like protobuf? We use JSON for CRDs because CRDs only support JSON

Reply:
I dont have any preference. Go's json stdlib is indeed slow-ish compared to other serialization format, but usually its not significant in practice i think.